integration/log_integration_test.go

```diff
@@ -276,6 +293,55 @@ func checkLogRootHashMatches(logID int64, tree *merkle.InMemoryMerkleTree, clien
 	return nil
 }
 
+// checkInclusionProofsAtIndex obtains and checks proofs at tree sizes from zero up to 2 x the sequencing
+// batch size (or number of leaves queued if less). The log should only serve proofs for indices in a tree
+// at least as big as the index where STHs where the index is a multiple of the sequencer batch size. All
+// proofs returned should match ones computed by the alternate Merkle Tree implementation, which differs
+// from what the log uses.
+func checkInclusionProofsAtIndex(testIndex int64, logID int64, tree *merkle.InMemoryMerkleTree, client trillian.TrillianLogClient, params testParameters) error {
+	for treeSize := int64(0); treeSize < min(params.leafCount, int64(2 * params.sequencerBatchSize)); treeSize++ {
+		ctx, cancelFunc := getRPCDeadlineContext()
+		resp, err := client.GetInclusionProof(ctx, &trillian.GetInclusionProofRequest{LogId: logID, LeafIndex:testIndex, TreeSize:int64(treeSize)})
+		cancelFunc()
+
+		if testIndex >= treeSize {
+			// Firstly, if the index is larger than the tree size we cannot have a valid proof
+			if err == nil {
+				return fmt.Errorf("log returned proof for index: %d, tree is only size %d", testIndex, treeSize)
+			}
+		} else {
+			// Secondly, if we're not at a valid STH tree size then we can't have a proof
+			if treeSize == 0 || (treeSize % int64(params.sequencerBatchSize)) != 0 {
+				if err == nil {
+					return fmt.Errorf("log returned proof at non STH size: %d", treeSize)
+				}
+			} else {
+				// Otherwise we should have a proof, to be compared against our memory tree
+				if err != nil || resp.Status.StatusCode != trillian.TrillianApiStatusCode_OK {
+					return fmt.Errorf("log returned no proof for index %d at size %d, which should have succeeded: %v", testIndex, treeSize, err)
+				}
+
+				// Remember that the in memory tree uses 1 based leaf indices
+				path := tree.PathToRootAtSnapshot(int(testIndex + 1), int(treeSize))
+
+				// Compare the proof lengths
+				if got, want := len(resp.Proof.GetProofNode()), len(path); got != want {
+					return fmt.Errorf("proof s:%d l:%d differs in length: %d %d", treeSize, testIndex, got, want)
+				}
+
+				// Then the node hashes should all match
+				for i := 0; i < len(path); i++ {
+					if got, want := base64.StdEncoding.EncodeToString(resp.Proof.GetProofNode()[i].NodeHash), base64.StdEncoding.EncodeToString(path[i].Value.Hash()); got != want {
```

Comment:
Not sure. There might be cases where it helps but it's really one thing doesn't match another and it's going to need investigation as to why.